async object detect

abdelaziz-mahdy opened the pull request:
can you check the code and let me know if there something i am doing wrong, i am getting many errors in dart file, i was going to fix those, but i want to be sure my impl of c++ is correct before moving on to dart fully

rainyl commented:
Ok, I will take a look.

rainyl commented on src/objdetect/objdetect_async.cpp:
```diff
@@ -0,0 +1,419 @@
+    END_WRAP
+}
+
+CvStatus *QRCodeDetector_Decode_Async(QRCodeDetector self, Mat input, CvCallback_2 callback) {
```
Here, the function will return 3 values, so a CvCallback_3 needed

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,665 @@
+
+  Future<bool> loadAsync(String name) async {
+    final cname = name.toNativeUtf8().cast<ffi.Char>();
+    final rval = cvRunAsync<bool>((callback) => CFFI.CascadeClassifier_Load_Async(ref, cname, callback), (c, p) {
```
Here the error is because no `CascadeClassifier_Load_Async` found, check it in `objdetect_async.h`, no declaration found.

abdelaziz-mahdy commented:
thank you for the mentioned fixes, i am not stuck in dart some function params i cant understand, and ffi size and the other size are confusing me. can you check and point me to the right direction?

after fixing dart file will make the test cases to make sure all is implemented and good to go

rainyl commented:
ffi.Size actually represents sizt_t, i.e., platform specific integers, I remember that ony several functions use it. so just ignore it.

cv::Size was wrapped with a struct in c header, and in dart it used to be typedef Size=(int, int), now it is wrapped with a class, but in order to keep compatibility, most functions were kept using (int, int).

Hope the above informations can help you. Also, please let me know if you have any suggestions.😄

abdelaziz-mahdy commented:
> ffi.Size actually represents sizt_t, i.e., platform specific integers, I remember that ony several functions use it. so just ignore it.
> 
> cv::Size was wrapped with a struct in c header, and in dart it used to be typedef Size=(int, int), now it is wrapped with a class, but in order to keep compatibility, most functions were kept using (int, int).
> 
> Hope the above informations can help you. Also, please let me know if you have any suggestions.😄

well i think thats the right way to do it, so all good, if i found a better way will let you know even though i think you did the best way

will wait for you to check the current dart impl, i always get lost with pointers and ref so sometimes as you can see i fail to do it correctly

rainyl commented:
I will check them later, but I am still considering how to organize async APIs, for functions they can be placed in a new file and it's fine, but for classes like HOGDescriptor, only one class name is allowed in dart, if sync and async are split to TWO files (like the objdetect_async.dart now), users have to import them separately.

So, the best solution I can think of is that, for functions, split to TWO files, but for classes, split every/several class to a single file and place sync and async in one class (one file), e.g., split HOGDescriptor to hogdescriptor.dart, and implement both sync and async inside it.

If you have better solutions, let's discuss it.

BTW, another solution is to split to different packages, like opencv_core, opencv_dnn, opencv_imgproc, etc., but it's laborious...

abdelaziz-mahdy commented:
> I will check them later, but I am still considering how to organize async APIs, for functions they can be placed in a new file and it's fine, but for classes like HOGDescriptor, only one class name is allowed in dart, if sync and async are split to TWO files (like the objdetect_async.dart now), users have to import them separately.
> 
> So, the best solution I can think of is that, for functions, split to TWO files, but for classes, split every/several class to a single file and place sync and async in one class (one file), e.g., split HOGDescriptor to hogdescriptor.dart, and implement both sync and async inside it.
> 
> If you have better solutions, let's discuss it.
> 
> BTW, another solution is to split to different packages, like opencv_core, opencv_dnn, opencv_imgproc, etc., but it's laborious...

different packages will be alot of work and i dont think its needed 

lets think about which will be easier for the users 
do you want them to be same class name? cant we do it like this 
```
      cascade_classifier.dart       // Sync version
      cascade_classifier_async.dart // Async version
```
with classes being 
```
class CascadeClassifier
class CascadeClassifierAsync
```

rainyl commented:
> i think the separation will make it easier to maintain and use, but the problem that sync and async cant be used in the same class, like if an instance got created async user cant use sync methods, is that a use case we should cover

I personally think we should, I will be inflexible if users have to import both sync and async version of the same API. so keep sync and async available in same class name is better.

> ```dart
> cascade_classifier.dart        // Main class
> cascade_classifier_sync.dart   // Sync methods
> cascade_classifier_async.dart  // Async methods
> ```

Oh, this looks better, I will test it in-depth.

Thanks for your insightful suggestions!

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,808 @@
+      int flags = 0,
+      (int, int) minSize = (0, 0),
+      (int, int) maxSize = (0, 0)}) async {
+    final ret = calloc<cvg.VecRect>();
+    final pnums = calloc<cvg.VecInt>();
+    final rval = cvRunAsync<(VecRect, VecInt)>(
+        (callback) => CFFI.CascadeClassifier_DetectMultiScale2_Async(
+            ref,
+            image.ref,
+            scaleFactor,
+            minNeighbors,
+            flags,
+            minSize.cvd.ref,
+            maxSize.cvd.ref,
+            ret,
+            pnums,
+            callback), (c, _) {
+      return c.complete((VecRect.fromPointer(ret), VecInt.fromPointer(pnums)));
+    });
+    return rval;
```
Here, please always remember in the async version, no need to calloc pointers at dart side, the callbacks will return them from native side
```suggestion
    // if the callback returns nothing (i.e., CvCallback_0()), use cvRunAsync0, if returns 1 value (i.e., CvCallback_1(void *)), use cvRunAsync1, if returns 2 values (i.e., CvCallback_2(void *, void *)), use cvRunAsync2, etc.
    final rval = cvRunAsync2<(VecRect, VecInt)>(
        (callback) => CFFI.CascadeClassifier_DetectMultiScale2_Async(
            ref,
            image.ref,
            scaleFactor,
            minNeighbors,
            flags,
            minSize.cvd.ref,
            maxSize.cvd.ref,
            callback),
            (c, p1, p2) {
                // the returned pointer are all ffi.Pointer<ffi.Void>, so we need to cast them to the correct type
                return c.complete((VecRect.fromPointer(p1.cast<cvg.VecRect>()), VecInt.fromPointer(p2.cast<cvg.VecInt>())));
    });
    return rval;
```

abdelaziz-mahdy replied:
Oh got, thank you, will recheck everything hopefully all goes well

rainyl replied:
yes, you are right

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,806 @@
+    return rval;
+  }
+
+    Future<(VecRect objects, VecInt numDetections, VecDouble levelWeights)> detectMultiScale3Async(
```
here numDetections should be rejectLevels, if the line is too long, just omit names, -> `Future<(VecRect, VecInt, VecDouble)>`

abdelaziz-mahdy replied:
I will search how to update the project formatter configuration, since removing names will make code harder to debug I think

abdelaziz-mahdy commented:
thank you for the example, i was able to fix most of the errors only one remains since it returns 4 params and there is no run_async4

abdelaziz-mahdy commented:
i added runasync4 and fixed the error, should i implement test cases or wait for you to have a look?

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+    (int, int) minSize = (0, 0),
+    (int, int) maxSize = (0, 0),
+  }) async {
+    final ret = calloc<cvg.VecRect>();
```
Here I think it's wrong, ret shouldn't be created in dart side

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+        (callback) => CFFI.CascadeClassifier_Empty_Async(ref, callback),
+        (c, p) {
+      final rval = p.cast<ffi.Bool>().value;
+      return c.complete(rval);
```
p should be freed or a memory leak will exists

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+        (callback) =>
+            CFFI.CascadeClassifier_getFeatureType_Async(ref, callback), (c, p) {
+      final rval = p.cast<ffi.Int>().value;
+      return c.complete(rval);
```
same as before, p should be freed explicitly

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+      calloc.free(p);
+      return c.complete(ret);
+    });
+    return rval;
```
Also, here p should be returned in onComplete rather than created in dart

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+            CFFI.CascadeClassifier_isOldFormatCascade_Async(ref, callback),
+        (c, p) {
+      final rval = p.cast<ffi.Bool>().value;
+      return c.complete(rval);
```
free

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+        (c, p) {
+      final rval = p.cast<ffi.Bool>().value;
+      calloc.free(cname);
+      return c.complete(rval);
```
free p

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+            ), (c, _, __) {
+      return c.complete(
+        (VecFloat.fromPointer(descriptors), VecPoint.fromPointer(locations)),
+      );
```
same as before, descriptors and locations shouldn't be created here, use the pointers provided by onComplete,
```dart
(c, p, p1) =>
      c.complete(
        (VecFloat.fromPointer(p.cast<cvg.VecFloat>()), VecPoint.fromPointer(p1.cast<cvg.VecPoint>())),
      )
```

abdelaziz-mahdy commented:
Should I free values coming from callbacks too?

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+      return c.complete((grad, angleOfs));
+    });
+    return rval;
+  }
```
https://docs.opencv.org/4.10.0/d5/d33/structcv_1_1HOGDescriptor.html#a1f76c51c08d69f2b8a0f079efc4bd093

grad and angleOfs are InputOutputArray, which means they should be included in parameters and provided by users

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,912 @@
+    double groupThreshold = 2.0,
+    bool useMeanshiftGrouping = false,
+  }) async {
+    final rects = calloc<cvg.VecRect>();
```
same as before

rainyl commented:
> Should I free values coming from callbacks too?

Yes, they are pointers, but only for primitive types like int, float, double, etc., for structs like Mat, Size, VecPoint etc., corresponding dart classes will be created using `.fromPointer() `constructor, so the created instance will manage the memory in later operations.

rainyl commented:
https://github.com/rainyl/opencv_dart/blob/b966c6eb0e9ee580a046bd9a99d312756c76b36f/lib/src/core/base.dart#L186-L202

You can refer to the impl in #117 , if only one value will be returned, we can simplify the onComplete by reusable functions like intCompleter, matCompleter, vecPointCompleter etc., I have moved matCompleter to mat.dart, vecPointCompleter to point.dart, etc., you can copy those files to use them.

abdelaziz-mahdy commented:
> https://github.com/rainyl/opencv_dart/blob/b966c6eb0e9ee580a046bd9a99d312756c76b36f/lib/src/core/base.dart#L186-L202
> 
> You can refer to the impl in #117 , if only one value will be returned, we can simplify the onComplete by reusable functions like intCompleter, matCompleter, vecPointCompleter etc., I have moved matCompleter to mat.dart, vecPointCompleter to point.dart, etc., you can copy those files to use them.

i didnt use those, i think we can migrate to it after this is working and test cases are implemented so when we migrate nothing breaks

abdelaziz-mahdy commented:
for line length this is the only fix i found https://github.com/dart-lang/dart_style/issues/918#issuecomment-2041518288

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,917 @@
+        (callback) => CFFI.CascadeClassifier_Load_Async(ref, cname, callback),
+        (c, p) {
+      final rval = p.cast<ffi.Int>().value != 0;
+      calloc.free(cname);
```
cname should not be freed in this closure, but at the end of the function.

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,917 @@
+    return rval;
+  }
+
+  Future<(VecRect rectList, VecDouble weights)> groupRectanglesAsync(
+    VecRect rectList,
+    VecDouble weights,
+    int groupThreshold,
+    double eps,
+  ) async {
+    final rval = cvRunAsync<(VecRect, VecDouble)>(
+        (callback) => CFFI.HOGDescriptor_groupRectangles_Async(
+              ref,
+              rectList.ref,
+              weights.ref,
+              groupThreshold,
+              eps,
+              callback,
+            ), (c, _) {
+      return c.complete((rectList, weights));
+    });
+    return rval;
+  }
```
Here, rectList and weights are input/output, so they will be modified internally, thus a CvCallback_0/cvRunAsync0 is needed

rainyl commented on src/objdetect/objdetect_async.cpp:
```diff
@@ -0,0 +1,419 @@
+    callback(new int(0));
+    END_WRAP
+}
+
```
The setSVMDetector has no return values, so CvCallback_0 is required

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,917 @@
+    return CascadeClassifier._(p);
+  }
+
+  factory CascadeClassifier.fromFile(String filename) {
+    final p = calloc<cvg.CascadeClassifier>();
+    final cp = filename.toNativeUtf8().cast<ffi.Char>();
+    cvRun(() => CFFI.CascadeClassifier_NewFromFile(cp, p));
+    calloc.free(cp);
+    return CascadeClassifier._(p);
+  }
```
Do you think that we need to async constructors? Loading the model is a heavy operation, if users load it in main thread, especially in Flutter, UI may be blocked.

abdelaziz-mahdy replied:
Yes I think so too, will check if it's doable and let you know

rainyl replied:
Yes, use static methods.

> they are identical so that is not a problem i think, also async constructors will be named async right?

Yes, use `static Future<CascadeClassifier> emptyAsync() async {}` will be clearer.

to split sync and async to different files, maybe some thing like https://github.com/rainyl/opencv_dart/issues/118#issuecomment-2194193568 will be better?

abdelaziz-mahdy replied:
I like that, will do it like the comment it clear and separated, so will do and the async constructors

rainyl commented:
> for line length this is the only fix i found [dart-lang/dart_style#918 (comment)](https://github.com/dart-lang/dart_style/issues/918#issuecomment-2041518288)

I really do not want to commit settings.json, according to this comment,

> https://github.com/dart-lang/dart_style/issues/918#issuecomment-2036687775

I have added format to pre-commit to format the line length to 110, now the format will run before `git commit`, you may need to add the formatted files and commit again, although kind of fussy, it's the best solution for now I think.

abdelaziz-mahdy commented:
> add precommit script

is the script ran automatically? since i dont think so, i had to run it manually

rainyl commented:
> > add precommit script
> 
> is the script ran automatically? since i dont think so, i had to run it manually

I forgot that git won't commit hooks, maybe add a github action to format code is better, will be resolved later.

abdelaziz-mahdy commented:
async constructors is done too

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,765 @@
+
+  Future<(String rval, Mat straightQRcode)> decodeCurvedAsync(
+    InputArray img,
+    VecPoint points, {
+    OutputArray? straightQRcode,
+  }) async {
```
here straightQRcode is not needed

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,765 @@
+  }
+
+  Future<(String rval, VecPoint points, Mat straightQRcode)> detectAndDecodeCurvedAsync(
+    InputArray img, {
+    VecPoint? points,
+    Mat? straightQRcode,
+  }) async {
```
here both points and straightQRcode should be deleted

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,765 @@
+  }
+
+  Future<(String ret, VecPoint points, Mat straightCode)> detectAndDecodeAsync(
+    InputArray img, {
+    VecPoint? points,
+    OutputArray? straightCode,
+  }) async {
```
same as before

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,765 @@
+  }
+
+  Future<(bool ret, VecPoint points)> detectAsync(
+    InputArray input, {
+    VecPoint? points,
+  }) async {
```
same

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,765 @@
+  }
+
+  Future<(String ret, VecPoint? points, Mat? straightCode)> decodeAsync(
+    InputArray img, {
+    VecPoint? points,
+    Mat? straightCode,
+  }) async {
```
ditto

rainyl commented on lib/src/objdetect/objdetect_async.dart:
```diff
@@ -0,0 +1,765 @@
+  }
+
+  Future<(bool ret, VecPoint points)> detectMultiAsync(
+    InputArray img, {
+    VecPoint? points,
+  }) async {
```
ditto

rainyl reviewed (commented):
Great! Almost there, I think it's time to write test :smile:

abdelaziz-mahdy commented:
should tests be separated or the same files test sync and async?

rainyl commented:
> should tests be separated or the same files test sync and async?

The separated way is better for maintenance, but I am just lazy so placed them in one file, do as what you like.

abdelaziz-mahdy commented:
I see you implemented the objdetect tests, so I should just implement the dnn one?

rainyl commented:
Yes, I just have some free time so finished it, you can implement the tests for dnn

abdelaziz-mahdy commented:
> Yes, I just have some free time so finished it, you can implement the tests for dnn

Ok, will do it, just a question why is the callback API not blocking and the normal calls blocking, even though both run the same c code instead of returns it calls the callback?

Can you explain it more to me, I see that flutter does it the same for java platform interface for async tasks too, but why it's not blocking 

Will appreciate if there is a reference

rainyl commented:
> Ok, will do it, just a question why is the callback API not blocking and the normal calls blocking, even though both run the same c code instead of returns it calls the callback?
> 
> Can you explain it more to me, I see that flutter does it the same for java platform interface for async tasks too, but why it's not blocking

It's basically the credits to [NativeCallable.listener](https://api.flutter.dev/flutter/dart-ffi/NativeCallable/NativeCallable.listener.html), this API provided by dart::ffi also uses SendPort and ReceivePort internally, and it ensures the callback will be called **in the future** at the **same** isolate where the callback was created, and the pointers will be passed to that isolate too, so we can consume the pointers in that isolate.

However, because the pointers are unsendable across isolates, so if a Mat is constructed inside an isolate, it still can't be sent to another isolate, e.g., the following is ERROR:

```dart
import 'dart:isolate';

import 'package:opencv_dart/opencv_dart.dart' as cv;

void main(List<String> args) async {
  final im = await Isolate.run<cv.Mat>(() async => cv.imreadAsync("test/images/circles.jpg"));

  print(im.shape);  // ERROR!
}
```

However, because the callback will be called in the future, so it won't block the current/main isolate, with async/await, main isolate can process other instructions, for Flutter, this will NOT block main/UI thread so the app is still interactive.

abdelaziz-mahdy commented:
wow, interested to see if it can use multicores or not, but in most cases async is much better due to its simple usage vs isolates, but will need to further test it 

i added dnn_test but mat operations needs to be migrated to async 
ones i need but not migrated 
```
minMaxLoc
```
```
convertTo
```

i guess that should be the next step migrating core.h and core.cpp to async

rainyl commented:
> i added dnn_test but mat operations needs to be migrated to async ones i need but not migrated
> 
> ```
> minMaxLoc
> ```
> 
> ```
> convertTo
> ```
> 
> i guess that should be the next step migrating core.h and core.cpp to async

Yes, they are next steps, for now just kepp using sync version.

abdelaziz-mahdy commented:
done, i am missing the models, so i cant test the code, dont you think a script to download all the needed models and running the tests is needed? also i workflow to make sure all tests are passing, since i cant find it (correct me if it does exist )

rainyl commented:
The models are located at https://github.com/rainyl/opencv_dart/releases/tag/dnn_test_files, the coverage workflow will download them, so don't worry~

BTW, I have fixed some string conversion issues, now I think it's time to merge

abdelaziz-mahdy commented:
Let me know what you plan to work on, so I work on something else, if you want to give me a module I don't mind 


I learned alot of ffi doing this so I am enjoying the learning experience and your explanation is awesome so thank you very much.

rainyl commented:
Let's open new issues to track and assign them.
